src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageActionDescriptorBuilder.cs

```diff
@@ -31,6 +32,15 @@ public static class CompiledPageActionDescriptorBuilder
                 .ToArray();
             var handlerMethods = CreateHandlerMethods(applicationModel);
 
+            if (applicationModel.ModelType != null && applicationModel.DeclaredModelType != null &&
+                !applicationModel.DeclaredModelType.IsAssignableFrom(applicationModel.ModelType))
+            {
+                var message = $"The model type for '{applicationModel.ActionDescriptor.DisplayName}' is of type " +
+                    $"'{applicationModel.ModelType.Name}' which is not assignable to its declared model type " +
+                    $"'{applicationModel.DeclaredModelType.Name}'.";
+                throw new InvalidOperationException(message);
```

Comment:
I'll put this inside a resource before checkin-in once you are ok with the wording.